internal/goro/adaptive_pool_test.go

```diff
@@ -0,0 +1,172 @@
+// The MIT License
+//
+// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package goro_test
+
+import (
+	"sync"
+	"testing"
+	"time"
+
+	"go.temporal.io/server/common/clock"
+	"go.temporal.io/server/internal/goro"
+)
+
+func block()   { <-make(chan struct{}) }
+func nothing() {}
+
+func TestAdaptivePool_CallsF(t *testing.T) {
+	t.Parallel()
+	ts := clock.NewEventTimeSource()
+
+	p := goro.NewAdaptivePool(ts, 1, 10, 10*time.Millisecond, 10)
+	defer p.Stop()
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	p.Do(wg.Done)
+	wg.Wait()
+}
+
+func TestAdaptivePool_Grows(t *testing.T) {
+	t.Parallel()
+	ts := clock.NewEventTimeSource()
+
+	p := goro.NewAdaptivePool(ts, 5, 10, 10*time.Millisecond, 10)
+	defer p.Stop()
+
+	// occupy five workers
+	p.Do(block)
+	p.Do(block)
+	p.Do(block)
+	p.Do(block)
+	p.Do(block)
+
+	// sixth call will still start new worker after delay
+	doneCh := make(chan struct{})
+	go func() {
+		p.Do(block)
+		doneCh <- struct{}{}
+	}()
+
+	// wait for goroutine to block in Do
+	time.Sleep(10 * time.Millisecond)
```

Comment:
Unfortunately the Sleeps are all required to make the synchronization work. The goroutine has to call NewTimer before we call Advance

Reply:
I think this is fine but if you really wanted 0 chance of flakiness you could control the firing of the timers since you've already injected a time source.

Reply:
I guess there are a few options.. if this was a same-package test instead of external, it could just reach in and check the worker count at various times. I thought I should try to do it based on just behavior though.

How would you control the firing of the timers? The things I thought of make the application code a lot more complicated just for the sake of the test.

Reply:
@dnr not true: https://github.com/temporalio/temporal/pull/6034/files#diff-179d947e11ceb58cd9c72d200727c6988a5e586eabc7fd35448d0201d7270d04R226

You can check the goroutine stacks to identify blocking

Reply:
I misread, I thought using `ts.Advance()` would work here but that's not good enough. You could instead of `block` have a function that updates the number of times it was called and wait for the expected number of calls here.

Reply:
Wow, that is... :astonished:  Honestly I'd rather make it a white-box test and read out the worker count. That seems a lot less fragile.

 I think changing block to a channel send would work well though. I'll try that.

Reply:
I managed to remove all but one sleep with a few tricks. I ran it 20000 times locally with no failures (though that doesn't mean much)